// reapp-kit/dist/index.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

var _Promise = require('bluebird');

// fetch

require('isomorphic-fetch');

var _UI = require('reapp-ui');

var _UI2 = _interopRequireWildcard(_UI);

var _Helpers = require('reapp-ui/helpers');

var _Helpers2 = _interopRequireWildcard(_Helpers);

require('reapp-object-assign');

// data

var _Immutable = require('immutable');

var _Immutable2 = _interopRequireWildcard(_Immutable);

var _store = require('./lib/store');

var _store2 = _interopRequireWildcard(_store);

var _Store = require('./lib/StoreComponent');

var _Store2 = _interopRequireWildcard(_Store);

var _action = require('./lib/action');

var _action2 = _interopRequireWildcard(_action);

// theme

var _theme = require('./lib/theme');

var _theme2 = _interopRequireWildcard(_theme);

// ui

var _Components = require('reapp-ui/all');

var _Components2 = _interopRequireWildcard(_Components);

// router

var _Router = require('react-router');

var _Router2 = _interopRequireWildcard(_Router);

var _generator = require('reapp-routes/react-router/generator');

var _generator2 = _interopRequireWildcard(_generator);

var _render = require('reapp-routes/react-router/render');

var _render2 = _interopRequireWildcard(_render);

// reapp

var _Theme = require('reapp-ui/helpers/Theme');

var _Theme2 = _interopRequireWildcard(_Theme);

var _Reapp = require('./lib/Reapp');

var _Reapp2 = _interopRequireWildcard(_Reapp);

// react patch

var _React = require('react');

var _React2 = _interopRequireWildcard(_React);

var _Component = require('./lib/Component');

var _Component2 = _interopRequireWildcard(_Component);

var _Page = require('./lib/Page');

var _Page2 = _interopRequireWildcard(_Page);

_Promise.Promise.longStackTraces();
window.Promise = _Promise.Promise;

_React2['default']._Component = _React2['default'].Component;
_React2['default'].Component = _Component2['default'];
_React2['default'].Page = _Page2['default'];

// inappbrowser
if (window.cordova && window.cordova.InAppBrowser) window.open = window.cordova.InAppBrowser.open;

module.exports = Object.assign(_Components2['default'], {
  // react
  React: _React2['default'],

  // reapp component/page export
  Component: _Component2['default'],
  Page: _Page2['default'],

  // immutable
  Immutable: _Immutable2['default'],

  // promise
  Promise: _Promise.Promise,

  // reapp
  Reapp: _Reapp2['default'],

  // data
  action: _action2['default'],
  store: _store2['default'],
  Store: _Store2['default'],

  // UI
  theme: _theme2['default'],
  makeStyles: _UI2['default'].makeStyles,
  Helpers: _Helpers2['default'],

  // router
  Router: _Router2['default'],
  router: function router(require, routes, opts, cb) {
    return _render2['default'].async(_generator2['default'].routes(require, routes), opts, cb);
  },
  route: _generator2['default'].route
});
