// reapp-routes/react-router/ParentRouteMixin.js
'use strict';

var React = require('react');
var assign = require('react/lib/Object.assign');

var REF_NAME = '__routeHandler__';

module.exports = {

  contextTypes: {
    routeDepth: React.PropTypes.number.isRequired,
    router: React.PropTypes.func.isRequired
  },

  childContextTypes: {
    routeDepth: React.PropTypes.number.isRequired
  },

  getChildContext: function getChildContext() {
    return {
      routeDepth: this.context.routeDepth + 1
    };
  },

  componentDidMount: function componentDidMount() {
    this._updateRouteComponent(this.refs[REF_NAME]);
  },

  componentDidUpdate: function componentDidUpdate() {
    this._updateRouteComponent(this.refs[REF_NAME]);
  },

  componentWillUnmount: function componentWillUnmount() {
    this._updateRouteComponent(null);
  },

  _updateRouteComponent: function _updateRouteComponent(component) {
    this.context.router.setRouteComponentAtDepth(this.getRouteDepth(), component);
  },

  getRouteDepth: function getRouteDepth() {
    return this.context.routeDepth;
  },

  createChildRouteHandler: function createChildRouteHandler(props) {
    var route = this.context.router.getRouteAtDepth(this.getRouteDepth());
    var el = route ? React.createElement(route.handler, assign({}, props || this.props, { ref: REF_NAME })) : null;
    return el;
  }

};
