// reapp-ui/components/Card.js
'use strict';

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Component = require('../component');

module.exports = Component({
  name: 'Card',

  propTypes: {
    title: React.PropTypes.string,
    children: React.PropTypes.node
  },

  getDefaultProps: function getDefaultProps() {
    return {
      animation: 'card'
    };
  },

  makeSection: function makeSection(title, content) {
    return content && React.createElement(
      'span',
      this.componentProps(title),
      content
    );
  },

  render: function render() {
    var _props = this.props;
    var children = _props.children;
    var title = _props.title;

    var props = _objectWithoutProperties(_props, ['children', 'title']);

    return React.createElement(
      'div',
      _extends({}, this.componentProps(), props),
      this.makeSection('title', title),
      this.makeSection('content', children)
    );
  }
});
