// reapp-ui/components/Chat.js
'use strict';

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Component = require('../component');
var clone = require('../lib/niceClone');
var ChatItem = require('./ChatItem');

var Chat = Component({
  name: 'Chat',

  propTypes: {
    // pass props to ChatItem
    itemProps: React.PropTypes.object,

    // auto wrap children with ChatItem
    wrap: React.PropTypes.bool
  },

  render: function render() {
    var _props = this.props;
    var children = _props.children;
    var itemProps = _props.itemProps;
    var wrap = _props.wrap;

    var props = _objectWithoutProperties(_props, ['children', 'itemProps', 'wrap']);

    return React.createElement(
      'div',
      _extends({}, this.componentProps(), props),
      React.Children.map(children, function (li, i) {
        if (wrap) return React.createElement(
          ChatItem,
          _extends({}, itemProps, {
            index: i,
            key: i }),
          li.content || li
        );

        return clone(li, { key: i, index: i });
      })
    );
  }
});

Chat.Item = ChatItem;

module.exports = Chat;
