// reapp-ui/components/Input.js
'use strict';

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Component = require('../component');
var Checkbox = require('../components/Checkbox');
var Radio = require('../components/Radio');
var Label = require('../components/Label');

module.exports = Component({
  name: 'Input',

  statics: {
    liNoPad: true
  },

  getDefaultProps: function getDefaultProps() {
    return {
      type: 'text'
    };
  },

  render: function render() {
    var _props = this.props;
    var label = _props.label;
    var labelProps = _props.labelProps;

    var props = _objectWithoutProperties(_props, ['label', 'labelProps']);

    var element,
        elementProps = {};

    switch (this.props.type) {
      case 'checkbox':
        element = Checkbox;
        break;
      case 'radio':
        element = Radio;
        break;
      default:
        element = 'input';
        elementProps = this.componentProps('input');
    }

    var input = React.createElement(element, Object.assign({}, elementProps, props));

    if (!label) {
      return input;
    }return React.createElement(
      Label,
      _extends({ title: label }, labelProps),
      input
    );
  }
});
