// reapp-ui/components/Label.js
'use strict';

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Component = require('../component');

module.exports = Component({
  name: 'Label',

  propTypes: {
    title: React.PropTypes.string
  },

  statics: {
    liNoPad: true
  },

  render: function render() {
    var _props = this.props;
    var title = _props.title;
    var children = _props.children;

    var props = _objectWithoutProperties(_props, ['title', 'children']);

    return React.createElement(
      'label',
      _extends({}, this.componentProps(), props),
      title && React.createElement(
        'span',
        this.componentProps('title'),
        title
      ),
      children
    );
  }
});
