// reapp-ui/themes/ios/constants/base.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var supportsHairline = require('./supportsHairline')();

var brandColor = '#307cff';
var light = supportsHairline ? '#d9d9dc' : '#ddd';
var mid = '#bbb';

// cordova
var device = window.device || {};
var version = parseInt(device.version || 0, 10);
var standalone = window.navigator.standalone;
var ios7 = device.platform === 'iOS' && version >= 7;
var ios8 = device.platform === 'iOS' && version >= 8;

exports['default'] = {
  hairline: supportsHairline,
  onePx: supportsHairline ? '0.5px' : '1px',

  // homescreen app
  standalone: standalone,

  // cordova
  device: device,
  ios7: ios7,
  ios8: ios8,

  // statusbar
  statusBarHeight: ios7 || standalone ? 20 : 0,

  black: '#000',
  white: '#fff',
  light: light,
  mid: mid,
  dark: '#8e8e93',

  brandColor: brandColor,
  brandColorInactive: '#a3a3a3',

  active: brandColor,
  inactive: mid,

  activeBG: brandColor,
  activeColor: '#fff',

  edgeWidth: 30
};
module.exports = exports['default'];
