// reapp-ui/index.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

var _Stylesheet = require('react-style');

var _Stylesheet2 = _interopRequireWildcard(_Stylesheet);

var _normalize$normalizeAll = require('./lib/normalizeStyles');

// until React 0.14, we emulate the context
require('./lib/contextPatch');

require('./lib/desktopTouch');
require('reapp-object-assign');

if (window._FORCE_ENABLE_RAF || window.process && process.env &&
// if in production and not disabling raf batching
process.env.NODE_ENV === 'production' && !process.env.DISABLE_RAF) require('reapp-raf-batching');

// Stores constants, animations and styles
//   see themes/ios/all.js for an example usage.

// Constants should be added before styles, they are
// passed into any styles file that returns a function.
// You can load multiple constants files, so if you'd
// like load the iOS theme, then override a few constants

// Styles are one big object, with each key mapping to a
// component/view's name. The values of those keys map
// to the refs within each component, and finally those
// values map to objects with styles.

// Styles compiled to ReactStyle objects on init.
// You can also load multiple styles objects, and order
// determines winner (last in applies).

// Finally, animations are an object. Their keys map to
// functions that take in (index, step, props).

var UI = module.exports = {
  styles: {},
  animations: {},
  constants: {},

  getTheme: function getTheme() {
    return {
      styles: this.styles,
      animations: this.animations,
      constants: this.constants
    };
  },

  // constants are order sensitive and are overwritten on object
  addConstants: function addConstants() {
    for (var _len = arguments.length, constants = Array(_len), _key = 0; _key < _len; _key++) {
      constants[_key] = arguments[_key];
    }

    constants.forEach(function (constant) {
      // allow functions, we pass in current constants
      if (typeof constant === 'function') constant = constant(UI.constants);

      Object.assign(UI.constants, constant);
    });
  },

  // simply unions a list of animations objects
  addAnimations: function addAnimations() {
    for (var _len2 = arguments.length, animations = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      animations[_key2] = arguments[_key2];
    }

    animations.forEach(function (animation) {
      Object.keys(animation).forEach(function (key) {
        UI.animations[key] = animation[key];
      });
    });
  },

  // styles are order sensitive and pushed onto array
  addStyles: function addStyles() {
    for (var _len3 = arguments.length, styles = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      styles[_key3] = arguments[_key3];
    }

    styles.forEach(function (style) {
      // style: { styles: { key: requireFunc }, (include: [] | exclude: []) }

      var _ref = style.styles ? style : { styles: style };

      var styles = _ref.styles;
      var include = _ref.include;
      var exclude = _ref.exclude;

      var requireFunc = styles.__requireFunc;
      delete styles.__requireFunc;
      var styleKeys = Object.keys(styles);

      if (include && exclude) throw new Error('Cannot define include and exclude');

      // include or exclude certain styles
      UI._addStyles(requireFunc, include && include.length ? styleKeys.filter(function (x) {
        return include.indexOf(x) !== -1;
      }) : exclude && exclude.length ? styleKeys.filter(function (x) {
        return exclude.indexOf(x) === -1;
      }) : styleKeys);
    });
  },

  // do the actual adding
  // styles: { name: requireFunc }
  _addStyles: function _addStyles(requireFunc, styles) {
    styles.forEach(function (key) {
      var style = requireFunc(key);
      if (typeof style === 'function') style = style(UI.constants);

      if (!UI.styles[key]) UI.styles[key] = _Stylesheet2['default'].create(_normalize$normalizeAll.normalizeAll(style));else Object.keys(style).forEach(function (ref) {
        UI.styles[key][ref] = [].concat(UI.styles[key][ref], _normalize$normalizeAll.normalize(style[ref]));
      });
    });
  },

  // we just store key: true and __requireFunc: function
  // so webpack can conditionally require styles later
  makeStyles: function makeStyles(requireFunc, components) {
    var styles = {};
    styles.__requireFunc = requireFunc;

    components.forEach(function (key) {
      styles[key] = true;
    });

    return styles;
  }
};
