// reapp-ui/behaviors/LeftNavBehavior.js
"use strict";

var LeftNavBehavior = {
  ALL_PARALLAX_FADE: {
    parent: {
      translate: function translate(parentWidth, scrollLeft) {
        return {
          x: parentWidth - 0.5 * scrollLeft
        };
      },
      rotate: function rotate() {
        return null;
      },
      opacity: function opacity(parentWidth, scrollLeft) {
        return 0.5 + 0.5 * (1 - scrollLeft / parentWidth);
      }
    },
    child: {
      translate: function translate(parentWidth, scrollLeft) {
        return { x: parentWidth - scrollLeft };
      },
      rotate: function rotate() {
        return null;
      },
      opacity: function opacity() {
        return null;
      }
    }
  },
  NORMAL: {
    parent: {
      translate: function translate(parentWidth, scrollLeft) {
        return {
          x: parentWidth - scrollLeft
        };
      },
      rotate: function rotate() {
        return null;
      },
      opacity: function opacity() {
        return null;
      }
    }
  }
};

module.exports = LeftNavBehavior;
