// reapp-ui/views/DottedViewList.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Component = require('../component');
var ViewListMixin = require('./ViewListMixin');
var Dots = require('../components/Dots');
var DottedViewListBehavior = require('../behaviors/DottedViewListBehavior');

module.exports = Component({
  name: 'DottedViewList',

  mixins: [ViewListMixin],

  getDefaultProps: function getDefaultProps() {
    return DottedViewListBehavior;
  },

  getInitialState: function getInitialState() {
    return Object.assign(this.getViewListInitialState(), {
      activeViewIndex: 0
    });
  },

  onViewEntered: function onViewEntered(index) {
    if (this.state.activeViewIndex !== index) this.setState({ activeViewIndex: index });
  },

  render: function render() {
    return React.createElement(
      'div',
      null,
      this.getViewList(),
      React.createElement(Dots, _extends({
        total: this.props.children.length || 0,
        active: this.state.activeViewIndex,
        styles: this.getStyles('dots')
      }, this.props.dotProps))
    );
  }
});
