// reapp-ui/mixins/Classed.js
'use strict';

var classnames = require('classnames');

// uses this.name as base

module.exports = {
  componentWillMount: function componentWillMount() {
    this._classSets = {};
    this._classRefs = {};
    this.setClasses(this.props);
  },

  componentDidUpdate: function componentDidUpdate(nextProps) {
    this.setClasses(nextProps);
  },

  setClasses: function setClasses() {
    this._classSets = {};

    if (this.props.className) this.addClass('self', this.props.className);
  },

  getClasses: function getClasses(ref) {
    var refClassSet = this._classSets[ref];
    refClassSet = refClassSet || {};
    refClassSet[this.getClassName(ref)] = true;
    return refClassSet;
  },

  getClassSet: function getClassSet(ref) {
    return classnames(this.getClasses(ref));
  },

  addClass: function addClass(ref, className, conditional) {
    // shorthands
    if (typeof className === 'boolean' && !!className) className = ref;else if (!className) {
      className = ref;
      ref = 'self';
    }

    this._classSets[ref] = this._classSets[ref] || {};
    this._classSets[ref][className] = true;
  },

  removeClass: function removeClass(ref, className) {
    ref = ref || 'self';
    this._classSets[ref][className] = false;
  },

  getClassName: function getClassName(ref) {
    if (this._classRefs[ref]) {
      return this._classRefs[ref];
    }this._classRefs[ref] = ref === 'self' ? this.name : '' + this.name + '--' + ref;
    return this._classRefs[ref];
  },

  getSelector: function getSelector(key) {
    return '.' + this.getClassName(key);
  }
};
